src/main/java/apoc/concurrent/Semaphores.java

```diff
@@ -0,0 +1,116 @@
+        this.log.info("successfully registered Pools for @Context");
+    }
+
+    @Override
+    public void init() throws Exception {
+        Iterators.stream(apocConfig.getKeys(ApocConfig.APOC_SEMAPHORE_PREFIX))
+                .filter(key -> !key.equals(ApocConfig.APOC_SEMAPHORE_DEFAULT_NAME))
+                .forEach(key -> {
+                    int permits = apocConfig.getInt(key, -1);
+                    String semaphoreName = key.substring(ApocConfig.APOC_SEMAPHORE_PREFIX.length() + 1);
+                    semaphoreMap.put(semaphoreName, new Semaphore(permits));
+                    log.debug("created semaphore %s with %d permits", key, permits);
+                });
+    }
+
+    public <T> Stream<T> withSemaphore(TerminationGuard terminationGuard, String semaphoreName, Supplier<Stream<T>> supplier) {
+        if (semaphoreName == null) {
+            return supplier.get();
+        } else {
+            Semaphore semaphore = getSemaphoreOrThrow(semaphoreName);
+            long now = System.currentTimeMillis();
+            boolean permitAcquired = false;
+            try {
+
+                // we can't simply use semaphore.acquire() here since it blocks and doesn't give us a chance for checking termination status
+                log.debug("about to acquire semaphore %s", semaphoreName);
+                while (!semaphore.tryAcquire(100, TimeUnit.MILLISECONDS)) {
+                    terminationGuard.check();
+                }
+                permitAcquired = true;
+                log.debug("acquired semaphore %s in %d millis", semaphoreName, System.currentTimeMillis()-now);
+
+                return supplier.get().onClose(() -> {
+                    semaphore.release();
+                    log.debug("released semaphore %s after %d millis", semaphoreName, System.currentTimeMillis()-now);
+                });
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (RuntimeException e) {
+                if (permitAcquired) {
+                    semaphore.release();
+                }
+                log.debug("released semaphore %s after %d millis due to %s", semaphoreName, System.currentTimeMillis()-now, e.getMessage());
+                throw e;
+            }
+        }
+    }
+
+    public int availablePermits(String semaphoreName) {
+        Semaphore semaphore = getSemaphoreOrThrow(semaphoreName);
+        return semaphore.availablePermits();
+    }
+
+    public void add(String semaphoreName, int permits) {
+        if (semaphoreMap.get(semaphoreName) != null) {
+            throw new IllegalArgumentException("already existing semaphore :" + semaphoreName + ". Use remove first");
+        }
+        semaphoreMap.put(semaphoreName, new Semaphore(permits));
+    }
+
+    public void remove(String semaphoreName) {
+        semaphoreMap.remove(semaphoreName);
```

Comment:
should this also throw on failure? so that typos are spotted and semaphores don't remain that shouldn't?

Reply:
we don't do anything with the blocked threads when we remove them?